File: merge_sort.py
def merge_sort(array):
    if (len(array) < 2):
        return array
    
    halfway = len(array) // 2

    left = array[:halfway]
    right = array[halfway:]

    return merge_sort_helper(merge_sort(left), merge_sort(right))

def merge_sort_helper(left, right):
    left_len = len(left)
    right_len = len(right)

    l = r = s = 0

    sorted_arr = [None] * (left_len + right_len)

    while l < left_len and r < right_len:
        if left[l] <= right[r]:
            sorted_arr[s] = left[l]
            l += 1
        else:
            sorted_arr[s] = right[r]
            r += 1
        s += 1
    
    while l < left_len:
        sorted_arr[s] = left[l]
        l += 1
        s += 1
    while r < right_len:
        sorted_arr[s] = right[r]
        r += 1
        s += 1
    return sorted_arr

File: Sorting/mergeSort.js
const mergeSort = (array) => {
    if (array.length < 2) {
		return array
    }
	let halfway = Math.floor(array.length / 2)
	
	let leftArr = array.slice(0, halfway)
	let rightArr = array.slice(halfway)
	
	return mergeSortHelper(mergeSort(leftArr), mergeSort(rightArr))
}

const mergeSortHelper = (left, right) => {
    let leftLen = left.length
	let rightLen = right.length
	let l = 0
	let r = 0
	let s = 0
	
	const sortedArr = new Array(leftLen + rightLen)
	
	while (l < leftLen && r < rightLen) {
		if (left[l] <= right[r]) {
			sortedArr[s++] = left[l++]
		} else {
			sortedArr[s++] = right[r++]
		}
    }
    while (l < leftLen) {
        sortedArr[s++] = left[l++]
    }
    while (r < rightLen) {
        sortedArr[s++] = right[r++]
    }
	return sortedArr
}